### PR TITLE
[Do not merge, partial change to discuss] Have gecko_string_cache use &Atom instead of BorrowedAtom

### PR DESCRIPTION
**Pull request by SimonSapin**
This is part of a larger change discussed at https://github.com/servo/servo/pull/12391#issuecomment-232183942, but I’d like to get feedback on this specifically first.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12548)

### CODE REVIEW

**Comment by SimonSapin** [comment]
The idea is to replace `BorrowedAtom<'a>` with `&'a WeakAtom`, given `pub struct WeakAtom(nsIAtom);`.

`WeakAtom` is not `Copy` or `Clone` and only ever used behind a `&WeakAtom` reference so that you can’t get it on the stack, and that reference itself is the `*mut nsIAtom` pointer.

`Atom` dereferences to `WeakAtom`, which is more in line with how `String` and `&str` work.

Using a `&T` reference is important to make things generic in the `selectors` crate since it uses a `HashMap` with atom keys. `HashMap` supports lookups with a key of a different type from the one stored, though the `std::borrow::Borrow` trait:

```rust
pub trait Borrow<Borrowed: ?Sized> {
    fn borrow<'a>(&'a self) -> &'a Borrowed;
}
```

Note that `&'a Borrowed` is returned, not `Borrowed<'a>` since higher-kinded type constructors are only possible in current rust through advanced traits trickery.

File: ports/geckolib/string_cache/lib.rs [hunk 1]
```diff
@@ -30,90 +33,88 @@ pub mod atom_macro;
...
 macro_rules! ns {
-    () => { $crate::Namespace(atom!("")) };
+    () => { atom!("") }
 }
 
-#[derive(PartialEq, Eq)]
-pub struct Atom(*mut nsIAtom);
-#[derive(PartialEq, Eq, Debug, Hash, Clone)]
-pub struct Namespace(pub Atom);
+pub type Namespace = Atom;
```
Comment (SimonSapin) [inline ports/geckolib/string_cache/lib.rs]: This was mostly to get things building more easily. I can re-introduce the newtype and add `WeakNamespace`.

**Comment by SimonSapin** [comment]
The part that I think is kinda fishy is `struct WeakAtom(nsIAtom);` and using `&self` as a raw pointer.

f? @bholley @emilio @heycam

**Comment by SimonSapin** [comment]
Full change set, if you want more context:

* https://github.com/servo/string-cache/pull/165
* https://github.com/servo/rust-selectors/compare/generic-atom_
* https://github.com/servo/servo/compare/selectors-generic-atom_

**Comment by SimonSapin** [comment]
@bholley To respond to your IRC question: `nsIAtom` is not unsized, and the reason not to use it directly is that its fields are public and we don’t want users to mess with them:

`ports/geckolib/gecko_bindings/structs_*.rs`

```rust
#[repr(C)]
#[derive(Debug, Copy)]
pub struct nsIAtom {
    pub _base: nsISupports,
    pub _bitfield_1: u32,
    pub mHash: u32,
    /**
   * WARNING! There is an invisible constraint on |mString|: the chars it
   * points to must belong to an nsStringBuffer. This is so that the
   * nsStringBuffer::FromData() calls above are valid.
   */
    pub mString: *mut ::std::os::raw::c_ushort,
}
```

**Comment by bholley** [comment]
I would prefer to add some bindgen magic to restrict those fields - see https://github.com/servo/rust-bindgen/pull/20#issuecomment-234601703.

**Comment by SimonSapin** [comment]
We can always change to that once bindgen supportis there.

**Comment by bholley** [comment]
@Manishearth is adding that support right now, so maybe we can just use nsIAtom and add the belt-and-suspenders privacy annotations shortly afterwards? Would be less overall churn.

**Comment by SimonSapin** [comment]
Changes in dependencies being merged and published, I’ve opened https://github.com/servo/servo/pull/12571. Let’s move this discussion there.

**Pull request closed without merging**